openff/evaluator/datasets/curation/components/filtering.py

```diff
@@ -831,21 +831,73 @@ def _find_smirks_matches(smiles_pattern, *smirks_patterns):
         list of str
             The matched smirks patterns.
         """
+        from rdkit import Chem
+
+        def _rdmol_from_smiles(smiles: str) -> Chem.Mol:
+            """
+            Create an RDKit molecule from a SMILES string.
+
+            Parameters
+            ----------
+            smiles: str
+                The SMILES string to convert.
+
+            Returns
+            -------
+            rdmol: rdkit.Chem.rdchem.Mol
+                The RDKit molecule.
+
+            Taken from https://github.com/openforcefield/openff-toolkit/blob/0.12.1/openff/toolkit/utils/rdkit_wrapper.py#L953
+            """
+            # TODO: This function does not handle SMILES parsing failures gracefully.
+            rdmol = Chem.MolFromSmiles(smiles, sanitize=False)
+
+            Chem.SanitizeMol(
+                rdmol,
+                Chem.SANITIZE_ALL
+                ^ Chem.SANITIZE_ADJUSTHS
+                ^ Chem.SANITIZE_SETAROMATICITY,
+            )
+            Chem.SetAromaticity(rdmol, Chem.AromaticityModel.AROMATICITY_MDL)
+            Chem.AssignStereochemistry(rdmol)
 
-        from openff.toolkit.topology import Molecule
+            Chem.AddHs(rdmol)
+
+            return rdmol
+
+        def _has_match(rdmol: Chem.Mol, smarts: str) -> bool:
+            """
+            Run substructure matching, returning only whether or not matches were found.
+
+            Parameters
+            ----------
+            rdmol: Chem.Mol
+                The RDKit molecule to match against.
+            smarts: str
+                The SMARTS pattern to match.
+
+            Returns
+            -------
+            bool
+                Whether or not a match(es) was found.
+
+            Taken from https://github.com/openforcefield/openff-toolkit/blob/0.12.1/openff/toolkit/utils/rdkit_wrapper.py#L2306
+            """
+            # TODO: This function does not handle SMILES parsing failures gracefully.
+            qmol = Chem.MolFromSmarts(smarts)
+
+            match_kwargs = dict(
+                uniquify=False, maxMatches=2**32 - 1, useChirality=True
+            )
+
+            return len(rdmol.GetSubstructMatches(qmol, **match_kwargs)) > 0
```

Comment:
Would using `GetSubstructMatch` work here? I think the only difference is that this method returns a single match and so breaks after finding the first instance of a match.

Reply:
Seems good - I don't know the RDKit API that well and didn't look here.

If you have one handy, could you share a script that uses this filter on a large dataset? You mentioned the 70k ThermoML compounds earlier. I should know my way around Evaluator better than this.

Reply:
Sure this script currently returns an empty dataset when it should contain a few records with properties of heavy water 
```python
import pandas
from openff.evaluator.datasets.curation.components import filtering, thermoml
from openff.evaluator.datasets.curation.workflow import CurationWorkflow, CurationWorkflowSchema

N_PROCESSES = 2

# download thermol and cache
thermoml_data_frame = thermoml.ImportThermoMLData.apply(
        pandas.DataFrame(),
        thermoml.ImportThermoMLDataSchema(cache_file_name="thermoml.csv"),
        N_PROCESSES,
    )

# filter the dataset to only keep heavy water
curation_schema = CurationWorkflowSchema(component_schemas=[filtering.FilterBySmirksSchema(smirks_to_include=["[2H]O[2H]"], allow_partial_inclusion=True)])

# pandas dataframe of properties
heavy_water_dataset = CurationWorkflow.apply(thermoml_data_frame, curation_schema, N_PROCESSES)

heavy_water_dataset.to_csv("heavy_water.csv")
```